### DISCUSSION TITLE
Why not just use deep-translator?

### DISCUSSION THREAD

**Discussion by nidhaloff** [johnbumgarner/wordhoard#16] [Announcements]
Hi, I stumbled across this library by chance and I can't help but notice that the code (or at least a good part of it) under `wordhoard/utilities/deep_translator.py` is clearly copy-pasted from the[ deep-translator library](https://github.com/nidhaloff/deep-translator). Like even the name of the file is the name of the library. Most of the code is exactly the same as in deep-translator, even variable names. Why not just use the library? it is open source anyway

Like, I understand that you added some extra stuff to it, but it would have been cool to just use the library as it is (as a dependency) or for example to give credit in your README or docs.

I think as open source developers, it would be really great if we support each other and the bare minimum in this case would be just to mention or give credits to someone for their work, because most of the time it is taken for granted. 

Congratulation on your library/work. Keep up the good work.
IMHO, It would be cool if you just use deep-translator as a third party lib because you will get the last updates, so more features, definitely bug fixes and most importantly support.

**Comment by johnbumgarner** [comment]
All the translator modules are names after the source.  

- wordhoard/utilities/deep_translator.py  
- wordhoard/utilities/mymemory_translator.py
- wordhoard/utilities/google_translator.py

Also my code uses the [DeepL Python Library](https://pypi.org/project/deepl) and not your library.  I also noted that your library also calls the DeepL service. 

I wrote the deep_translator module code back on September 24, 2021.   Please let me know where you think I plagiarized your library in my code?

**Reply by nidhaloff** [reply]
So you are telling me that [this exception](https://github.com/johnbumgarner/wordhoard/blob/master/wordhoard/utilities/exceptions.py) file from your lib and [this](https://github.com/nidhaloff/deep-translator/blob/master/deep_translator/exceptions.py) is a coincidence? :D like the doc string is even the same... I remember writing the docstring on my own, so somehow we both had the same phrases to write? :D some of the docstrings are even mentioning deep_translator referring to my library.

Well, I started the library in June 2020. The variable names are the same, even the exception class names and exception messages are the same. The way your code is structured is also the same.

Like, I have no doubt that you took my code and just changed it a bit. Denying it is not cool!

I'm not here to argue, though. I just couldn't help but notice that and wanted to give you my 2 cents about it. At the end, It's up to you how to understand it.

**Comment by johnbumgarner** [comment]
Thanks for the info.  

I just compared my modules to your's and I don't see many similarities, but I will do a deep comparison with a DIFF tool.    My code structure was developed through years of coding and with help from friends that reviewed and tested the code prior to release.  But I will inspect my code for any similarities to yours.  I currently only plan to use DeepL, because that module has an API and the other modules make a request to a web link for the service being used.  I will review your Library to see if it is a good fit to replace the 3 translation modules in my current code base.   

Thanks again for raising this issue with me.

**Comment by johnbumgarner** [comment]
I checked these WordHoard modules for code similarities to your deep_translator modules.  There were no similarities, beyond the module names. 

- wordhoard/utilities/deep_translator.py
- wordhoard/utilities/mymemory_translator.py
- wordhoard/utilities/google_translator.py

I did check my exceptions class against yours and there were similarities.  I checked my development environment and I don't have deep_translator installed.  

I'm going to assume that when I was developing my code 2 years ago I found the exceptions class somewhere via my google searches for adding translation services to my code.  I apologize for not checking the source or referencing the source of that module's code.  I will resigned my code to fix these code similarities in the exceptions module. 

Additionally, I will review your deep_translator package to see if it is good fit for my WordHoard package.  It might be a good fit for another project that I'm working on where entire blocks of text might need to be translated by an end user.  My biggest fear here would be translation context, which is beyond the scope of your package.  

I would like to apologize again for these similarities in my exceptions class.

**Comment by nidhaloff** [comment]
I think we both know that it is more than just similarities! I mean, even the comments are the same...

Furthermore, I didn't claim that your whole package is like mine. So, running a DIFF tool on the whole package/module proves nothing, because only parts of my code have been copied, and clearly not everything. 

Basically, that exception file is the exact same. Also, your google_translate and memory_translate.py have some parts of my code, not everything, but some noticeable parts that required much research from my side. 

Now I would understand if you would say that you (or maybe a contributor) took inspiration from my work, which is totally fine, but at least give other devs credit for their work.

You don't have to use the package if it does not suit your use-case, but mentioning it in your README as a related work or inspiration would have been cool.

**Comment by johnbumgarner** [comment]
As I previously stated the exceptions class has some limited sections of your code, which I will rework.  

Again I have no idea where I got the exception class 2 years ago.  I will rework my code to ensure that there is no related code between my package and yours.   

I also have NEVER used your package in any of my projects.  

I will push an update package out soon and remove any similarities.

**Reply by nidhaloff** [reply]
This is not the point, though. Renaming the code you copied is as unfair as copying it without giving credits in the first place. 

You got me wrong. I don't mind taking parts of my code, it just feels unfair if you write your name as the author and give yourself copyrights for it. That's the problem. 
I'm not against it if done right, though. My code is open source anyway.

**Comment by johnbumgarner** [comment]
I redesigned my entire exception class, so there is no code related to your exception class in mine.  I also scrubbed the 3 WordHoard's translation modules for any code similarities related to the verbiage in several exceptions.

The next release of WordHoard will be pushed with these code updates after some testing.  

I did note that some of the exception class names used by both of us are also used by DeepL and other language translation packages.  

Reworking my code also allowed me to redesign and test the old code base for any bugs or performance issues.  I was also able to update my code with the latest python packages related to WordHoard's dependencies.

As I previously stated there will zero code similarities been your code and WordHoard v1.5.2 (next release).

I would like to apologize again for reusing a very small section of exception code from your code base without provided you attribution.

**Comment by johnbumgarner** [comment]
This discussion is being locked, because I corrected your concerns in WordHoard v1.5.2, which was just pushed to Github and will be pushed to PyPI today.
